### PR TITLE
Feature/tech 2/phrase matching feature only

### LINKED ISSUE #29: Add Support for Phrase Matching
# User Story

> As a user, I want to be able to enter a phrase in quotes and only see results that match that phrase, e.g., "cool beans" should only match "cool" followed by "beans" rather than each word individually (as is the current default).

# Acceptance Criteria

- Add a new method to the controller called `enable_phrase_matching` that, when called as part of configuration, causes phrase matching to be turned on for all queries that are run through the search implementation.

# Implementation Details

Example of how this could be used in code:

```php
add_action(
	'elasticsearch_extensions_config',
	 function( $es_config ) {
		$es_config->enable_phrase_matching();
	 }
);
```

- When enabled, search queries should be evaluated for text in double quotes, which should be included in the search as a `match_phrase`.
- Any other text in the search, not in quotes, should be searched for normally.
- The solution should support multiple phrase searches in the same query (e.g., `"quick brown" fox "jumped over"`).
- This feature should be opt-in and should be disabled by default.
- There should be a corresponding `disable_phrase_matching` function to turn this feature off, which could be used by plugins or themes conditionally to unset this setting in particular contexts (otherwise there would be no way to turn it off once it was turned on).

### PR DESCRIPTION
**Pull request by jakewrfoster**
## Overview

Adds feature for phrase matching to ESX plugin which is configurable via the `'elasticsearch_extensions_config'` hook. This PR intentionally does not include tests as they are being worked on in #64. This PR just includes the feature itself.

## Closes

#29 

## Changes

* Adds a new method to the controller called `enable_phrase_matching` that, when called as part of configuration, causes phrase matching to be turned on for all queries that are run through the search implementation. i.e.

```php
add_action(
	'elasticsearch_extensions_config',
	 function( $es_config ) {
		$es_config->enable_phrase_matching();
	 }
);
```

* When enabled, search queries are evaluated for text in double quotes (delineator `"` is filterable), which is included in the search as a multimatch clause.
* Any other text in the search, not in quotes (or the filtered delineator), is searched for normally.
* Multiple phrase searches in the same query (e.g., "quick brown" fox "jumped over") are supported.
* This feature is opt-in and is disabled by default.
* There is a corresponding `disable_phrase_matching` function to turn this feature off, which could be used by plugins or themes conditionally to unset this setting in particular contexts (otherwise there would be no way to turn it off once it was turned on).

## Summary by CodeRabbit

- **New Features**
  - Introduced phrase matching capabilities for enhanced search functionality.
  - Added methods to enable or disable phrase matching in search queries.
  - Implemented aggregation addition to Elasticsearch query formatting.

- **Refactor**
  - Refactored the logic for adding aggregations to Elasticsearch queries.

- **Tests**
  - Added test cases for the new phrase matching feature.

- **Chores**
  - Cleaned up and organized code related to search query argument handling.

### CODE REVIEW

File: lib/adapters/class-adapter.php
Comment (renatonascalves) [inline lib/adapters/class-adapter.php:45]: Is it worth adding the link from the es documentation? So that anyone looking in can better understand what es feature this implements?

File: lib/adapters/class-adapter.php
Comment (renatonascalves) [inline lib/adapters/class-adapter.php:325]: Is it intentional that only search global queries is handled? I'm thinking of custom requests with `WP_Query` or the REST API where this query var is not set globally.

File: lib/adapters/class-adapter.php
Comment (renatonascalves) [inline lib/adapters/class-adapter.php:333]: Missing filter hook.

File: lib/adapters/class-adapter.php
Comment (renatonascalves) [inline lib/adapters/class-adapter.php:366]: Ideally, this could, or should, respect the active provider. One could easily hook into [VIP Search](https://github.com/Automattic/vip-go-mu-plugins-built/blob/93d4ba09a0fd308185867905a593edebe748290f/search/elasticpress/includes/classes/Indexable/Post/Post.php#L1459) and set a list of custom fields and forget, or not know, about this hook and have two, or more, separate fields for each query in a single request.

Comment (renatonascalves) [reply lib/adapters/class-adapter.php:366]: The fields used here also have weights assigned to them. That to me assumes the default behavior has the same behavior. This can skew the results.

File: lib/adapters/class-adapter.php
Comment (renatonascalves) [inline lib/adapters/class-adapter.php:336]: Is it worth checking for `$matches[1]` instead of `$matches`?

File: lib/adapters/class-adapter.php
Comment (renatonascalves) [inline lib/adapters/class-adapter.php:317]:
```suggestion
	 * @param array $es_args Elasticsearch DSL to filter.
```

File: lib/adapters/class-adapter.php
Comment (renatonascalves) [inline lib/adapters/class-adapter.php:669]:
```suggestion
	public function set_enable_phrase_matching( bool $enable_phrase_matching ): void {
```

File: lib/adapters/class-vip-enterprise-search.php
Comment (renatonascalves) [inline lib/adapters/class-vip-enterprise-search.php:39]:
```suggestion
	public function add_aggs_to_es_query( $formatted_args ): array {
```

File: lib/adapters/class-vip-enterprise-search.php
Comment (renatonascalves) [inline lib/adapters/class-vip-enterprise-search.php:36]:
```suggestion
	 * @param array $formatted_args Elasticsearch DSL.
```

File: lib/adapters/class-vip-enterprise-search.php
Comment (renatonascalves) [inline lib/adapters/class-vip-enterprise-search.php:236]:
```suggestion
	public function filter__ep_post_formatted_args( $formatted_args, $args ): array {
```

File: lib/class-controller.php
Comment (renatonascalves) [inline lib/class-controller.php:58]:
```suggestion
	 * Disables phrase matching for search queries.
```

**Review by renatonascalves — CHANGES_REQUESTED** [review]
Left a few comments. It'd be great to have the [wiki](https://github.com/alleyinteractive/elasticsearch-extensions/wiki) updated with this new addition after this is merged.

File: lib/adapters/class-adapter.php
Comment (jakewrfoster) [reply lib/adapters/class-adapter.php:333]: I don't want to block the feature by this filter, I will add this as an issue/enhancement!

File: lib/adapters/class-adapter.php
Comment (jakewrfoster) [reply lib/adapters/class-adapter.php:336]: if `$matches` isn't empty then `$matches[1]` is implied (since there has to be at least one match for `$matches` to be defined). At any rate, it is good to be explicit, so I will update it.

File: lib/adapters/class-adapter.php
Comment (jakewrfoster) [reply lib/adapters/class-adapter.php:366]: These felt like sane defaults to me. This is another place I think where we could add an enhancement issue to hook into the provider and the already filtered values for these fields. Starting with a filterable set of fields to me, feels like a good first step.

File: lib/adapters/class-adapter.php
Comment (jakewrfoster) [reply lib/adapters/class-adapter.php:325]: This is intentional as per the use cases that I have been asked to cater to so far.

File: lib/adapters/class-adapter.php
Comment (jakewrfoster) [reply lib/adapters/class-adapter.php:45]: I can add some links!

File: lib/adapters/class-adapter.php
Comment (jakewrfoster) [reply lib/adapters/class-adapter.php:333]: https://github.com/alleyinteractive/elasticsearch-extensions/issues/68

File: lib/adapters/class-vip-enterprise-search.php
Comment (jakewrfoster) [reply lib/adapters/class-vip-enterprise-search.php:36]: this is an array of formatted  ES args, it filters `ep_post_formatted_args`

File: lib/adapters/class-adapter.php
Comment (renatonascalves) [reply lib/adapters/class-adapter.php:334]: I see! Is it possible to document this somehow in the wiki?

File: lib/adapters/class-adapter.php
Comment (renatonascalves) [reply lib/adapters/class-adapter.php:375]: If that's documented, sure. If not, I'll realize, or not, later after activating it that my search results are not behaving as quite expected.

File: lib/adapters/class-adapter.php
Comment (jakewrfoster) [reply lib/adapters/class-adapter.php:334]: I will add a wiki entry as soon as this is merged in :)

File: lib/adapters/class-adapter.php
Comment (jakewrfoster) [reply lib/adapters/class-adapter.php:375]: these DSL changes are only applied if the delineators are found in the query, for example a search for `brown dog` will not include a multi_match phrase query in the DSL. On the other hand, a search for `"brown dog"` will return results that only have the exact phrase in one of these fields (weighted as they are above).

Comment (renatonascalves) [reply lib/adapters/class-adapter.php:375]: Correct! I think I got this part. The part I'm bringing attention to is that it will return results using fields and weights differently from the ones set by the site. Will it matter? Probably not, but due to the impact in the score, if you don't know this behavior is there, it does matter.

Comment (jakewrfoster) [reply lib/adapters/class-adapter.php:375]: I got ya :) i definitely agree that this filter should be hooked into from the adapters and from there the fields should be passed through from the providers. I also think that this should be added as an issue instead of blocking this PR, since it will be obvious that the feature is responsible for manipulating the query when the feature is being used (i.e. when double quotes are added to the search and the feature is turned on).

**Review by renatonascalves — APPROVED** [review]

File: lib/adapters/class-adapter.php
Comment (kevinfodness) [inline lib/adapters/class-adapter.php:197]: imo we don't need to filter it, double quote is standard for this. Also, you can check the results of `preg_match_all` rather than looking at the first item being empty, since it will return `0` on no match or `false` on failure, and a falsey return value from that function means no matches and you can bail.
```suggestion
		$phrase_match_delineator = '"';
		if ( ! preg_match_all( '/' . $phrase_match_delineator . '(.*?)' . $phrase_match_delineator . '/', $search, $matches ) ) {
			return $es_args;
		}
```

**Review by kevinfodness — APPROVED** [review]
One suggestion, otherwise 🍣

File: lib/adapters/class-adapter.php
Comment (jakewrfoster) [reply lib/adapters/class-adapter.php:197]: I will pull the filter out!

**Pull request merged**
